[CI:DOCS] man pages: document some --format options

edsantiago opened the pull request:
This is a resurrection of #14386. I know this will upset some people. I keep thinking about this problem, and am more committed than before to fixing it.

I believe #14046 is important. I believe we should (1) document the values that `--format` accepts, and (2) implement tests to make sure that our documentation doesn't stray from reality.

Defense: look at this PR, in particular, the two commits in it. First commit is #14386 as it was in May, second commit is what I've had to do because `podman pod inspect` and `podman stats` have added new values that I find super confusing and I actually think are wrong: `.InspectPodData` and `.ContainerStats` seem to add no value, they're just a level of indirection. I think a user playing with command completion will find them confusing.

My purpose in this PR is to incrementally fix our documentation, with the end goal of making #14046 a super-trivla script-change-only PR that is easy to review. That can only be done by fixing the man pages a little at a time.

Unfortunately, I cannot do that. I don't have the knowledge or expertise to know what these fields mean. Which means I need allies. Can I please get help documenting these fields, and/or hiding the ones that should not be documented?

```release-note
Document recognized fields in `--format`, for some commands
```

rhatdan commented:
LGTM
I am fine with this change.

edsantiago commented:
...without the WTFs and FIXMEs, I assume...

rhatdan commented:
Sadly I missed those.

edsantiago commented:
Marking as draft to make sure this doesn't merge before I get my requested help

rhatdan commented on docs/source/markdown/podman-container-inspect.1.md:
```diff
@@ -18,6 +18,48 @@ all results in a JSON array. If a format is specified, the given template will b
+| .RestartCount        | Number of times container has been restarted (int) |
+| .Rootfs              | Container rootfs (string)                          |
+| .SizeRootFs          | Size of rootfs (int)                               |
+| .SizeRw              | FIXME                                              |
```
I believe this is the size of upper layer of the container, IE The modifications of the container, and the content shown when you run podman diff.

edsantiago replied:
Aha, got it. This field is only usable under `podman container inspect --size`. Without `--size`, it always returns `<nil>` (the string anglebracket, n-i-l, anglebracket). That seems worth documenting. I wonder if there's a way to issue a warning if someone tries to use it without `--size`?

mheon replied:
Shouldn't it imply `--size`? I swear we've had this discussion before in another issue...

edsantiago replied:
#4744 (and its PR, #4753) is the only instance I can find of `SizeRw` in a PR or issue. Of course, github search isn't showing me **this** PR when I search, so humph.

rhatdan commented on docs/source/markdown/podman-pod-inspect.1.md:
```diff
@@ -18,22 +18,42 @@ Change the default output format.  This can be of a supported type like 'json'
+| .Devices             | Devices                                     |
+| .ExitPolicy          | Exit policy                                 |
+| .InfraConfig ...     | Infra config (contains further fields)      |
+| .InspectPodData ...  | FIXME this is a deep nested confusing thing |
```
@cdoern PTAL ^^

rhatdan commented on docs/source/markdown/podman-stats.1.md:
```diff
@@ -30,17 +30,35 @@ Pretty-print container statistics to JSON or using a Go template
+| .ContainerStats ... | WTF WTF WTF 2022-09-12 this is a total dup dup?          |
+| .SystemNano         | Current system datetime, nanoseconds since epoch         |
+| .UpTime             | WTF? This is Duration (CPUNano) in human-readable form?? |
+| .Up                 | Same as UpTime (FIXME: why?)                             |
```
Could be for docker compatibility?
func (s *containerStats) Up() string {
	return (s.UpTime.String())
}

rhatdan replied:
Nope, I don't see this working on Docker. git blame blams @cdoern ^^

TomSweeneyRedHat replied:
Don'tcha just hate it when Git "blams" someone?  It's always a messy clean-up afterward....

cdoern replied:
where is the git blame on this one? This might've been from one of my first PRs...

rhatdan replied:
If we can just remove the Up function it might make this one go away.

rhatdan commented on docs/source/markdown/podman-stats.1.md:
```diff
@@ -30,17 +30,35 @@ Pretty-print container statistics to JSON or using a Go template
+| .PerCPU             | FIXME someone help please                                |
+| .ContainerStats ... | WTF WTF WTF 2022-09-12 this is a total dup dup?          |
+| .SystemNano         | Current system datetime, nanoseconds since epoch         |
+| .UpTime             | WTF? This is Duration (CPUNano) in human-readable form?? |
```
This is documented as the `CPU TIME`, Time since container started, Looks like a semi human form.

rhatdan replied:
Labeled as "CPU Time"

Looks like a time duration printed out as string.

rhatdan commented on docs/source/markdown/podman-stats.1.md:
```diff
@@ -30,17 +30,35 @@ Pretty-print container statistics to JSON or using a Go template
+| .BlockOutput        | Block Output                                             |
+| .PIDs               | Number of PIDs                                           |
+| .PerCPU             | FIXME someone help please                                |
+| .ContainerStats ... | WTF WTF WTF 2022-09-12 this is a total dup dup?          |
```
Strange this looks like a leak of the containerStats structure within the command.
@Luap99 @baude thoughts?

TomSweeneyRedHat commented on docs/source/markdown/podman-stats.1.md:
```diff
@@ -30,17 +30,35 @@ Pretty-print container statistics to JSON or using a Go template
+| .PerCPU             | FIXME someone help please                                |
+| .ContainerStats ... | WTF WTF WTF 2022-09-12 this is a total dup dup?          |
+| .SystemNano         | Current system datetime, nanoseconds since epoch         |
+| .UpTime             | WTF? This is Duration (CPUNano) in human-readable form?? |
```
If possible, it would be nice to have this list in alpha order.

edsantiago replied:
Oh boy, that one hasn't cost me sleep, but has cost me a lot of wasted time thinking about it. I settled on alpha-ish, in which important fields (ID, Name) are at top, and identical ones (ID & ContainerID, CreateCgroup & CgroupParent) are next to each other. But I'm glad you started the discussion, and welcome thoughts on what the least-worst approach is.

rhatdan replied:
I would guess the ContainerStats might be the problem.

@edsantiago You could break this into default ones and then additional ones.

edsantiago replied:
Like, two tables, one "You probably want..." and one "...but there's also"? That's doable, and seems user-friendly to me. I don't really have the judgment to know what's what, though.

And, thinking about it, now that I know how to put comments in a `.md` file ... if it's absolutely necessary to have fields that are spit out by command-completion but which must not be documented, we could (maybe) hide those fields in the markdown, in a way that the xref script will see them but human readers of the man page won't. (Users would still see those when they hit TAB though; they just wouldn't be in the man page). I don't especially care for this idea, but sometimes lesser evil etc.

TomSweeneyRedHat commented:
One small request.  Otherwise, LGTM once the fixmes are run to ground.

rhatdan commented on docs/source/markdown/podman-stats.1.md:
```diff
@@ -30,17 +30,35 @@ Pretty-print container statistics to JSON or using a Go template
+| .BlockInput         | Block Input                                              |
+| .BlockOutput        | Block Output                                             |
+| .PIDs               | Number of PIDs                                           |
+| .PerCPU             | FIXME someone help please                                |
```
Digging into cgroup I believe this is 
`cpuacct.usage_percpu: reports the CPU time (in nanoseconds) consumed on each CPU by all tasks in this cgroup (including tasks lower in the hierarchy). `

Looks like this only works in CgroupV1?
@giuseppe PTAL

edsantiago replied:
>Looks like this only works in CgroupV1?

Thanks, that could explain why all I get is `[]` (bracket-bracket) when I test it.

cdoern commented:
found my PR from a year and a little back, I can take a look at some of the stats issues here https://github.com/containers/podman/commit/f26fa5392110c2b59cb17d4f1013cf6bb4ecd873

Luap99 commented:
OK because I see the question, why are some fields redundant/duplicates? The answer is simple because we use embedded structs.
Examples from the completion [test](https://github.com/containers/podman/blob/main/cmd/podman/common/completion_test.go) :
```
type Anonymous struct {
	Hello string
	// The name should match the testStruct Name below. This is used to make
	// sure the logic uses the actual struct fields before the embedded ones.
	Name struct {
		Suffix string
		Prefix string
	}
}

testStruct := struct {
		Name string
		Age  int
		Car  *Car
		Car2 *Car
		*Anonymous
		private int
	}{}
```

Her Anonymous is embedded in the testStruct, that means it has no extra field name and therefore the Anonymous fields can be accessed directly on testStruct `.Hello` will work but you also can do `.Anonymous.Hello`

Now why should we suggest the same twice?
The problem comes when there are fields with the same name. We have `testStruct.Name` and  `testStruct.Anonymous.Name` both are fields with different information. So as a user I still need `testStruct.Anonymous.Name` if I need that specific field.

I hope that explains why we have duplicates in the suggestions.

Also sometimes we have duplicates for docker compat.

edsantiago commented:
Rebased and repushed with:
* squashed commit
* removed FIXMEs and WTFs
* added footnotes about options that require `--size` or CGroupsV1
* handwaved a few as "for experts only"
* acknowledged that some are dups
* sorted alphabetically, and added a check to my (so far private) xref tool

Basically, this ain't something I can do myself. All I want is to get things started, with bare-minimum descriptions, where the important thing is listing the format specifiers in the man page. I will follow up with other such PRs, adding skeleton descriptions to other man pages. Once (if) I get them all done, I will submit my xref script for inclusion in CI. Then, appearance/disappearance of any format specifiers will trigger a CI failure unless the change is reflected in man pages. See #14046.

rhatdan commented:
SGTM

rhatdan commented:
Still need to remove Draft status.

rhatdan commented:
/lgtm
